src/components/manager.ts

```diff
@@ -1,75 +1,82 @@
 import * as utils from '../utils'
 
 import {Extension} from '../main'
 
+interface Content {
+    [id: string]: { // tex file name
+        content: string, // the dirty (under editing) contents
+        children: { // sub-files, should be tex or plain files
+            index: number, // the index of character sub-content is inserted
+            file: string // the path to the sub-file
+        }[]
+    }
+}
 
 export class Manager {
-    extension: Extension
-    rootFiles: { [key: string]: string }
-    localRootFiles: { [key: string]: string | undefined }
-    workspace: string
-    texFileTree: { [id: string]: Set<string> } = {}
-    fileWatcher: chokidar.FSWatcher
-    bibWatcher: chokidar.FSWatcher
-    filesWatched: string[]
-    bibsWatched: string[]
-    watcherOptions: chokidar.WatchOptions = {
+    cachedContent: Content = {}
+
+    private extension: Extension
+    private fileWatcher: chokidar.FSWatcher
+    private bibWatcher: chokidar.FSWatcher
+    private filesWatched: string[] = []
+    private bibsWatched: string[] = []
+    private watcherOptions: chokidar.WatchOptions = {
         usePolling: true,
         interval: 300,
         binaryInterval: 1000
     }
 
     constructor(extension: Extension) {
         this.extension = extension
-        this.filesWatched = []
-        this.bibsWatched = []
-        this.rootFiles = {}
-        this.localRootFiles = {}
-        this.workspace = ''
     }
 
-    getOutputDir(texPath: string) {
+    /* Returns the output directory developed according to the input tex path
+       and 'latex.outDir' config. If undefined is passed in, the default root
+       file is used. */
+    getOutDir(texPath?: string) {
+        if (texPath === undefined) {
+            texPath = this.rootFile
+        }
         const doc = texPath.replace(/\.tex$/, '').split(path.sep).join('/')
         const docfile = path.basename(texPath, '.tex')
         const configuration = vscode.workspace.getConfiguration('latex-workshop')
         const docker = configuration.get('docker.enabled')
-        let outDir = (configuration.get('latex.outDir') as string)
-        outDir = outDir.replace(/%DOC%/g, docker ? docfile : doc)
-                    .replace(/%DOCFILE%/g, docfile)
-                    .replace(/%DIR%/g, docker ? './' : path.dirname(texPath).split(path.sep).join('/'))
-                    .replace(/%TMPDIR%/g, this.extension.builder.tmpDir)
-        return outDir
+        let out = (configuration.get('latex.outDir') as string)
+        return out.replace(/%DOC%/g, docker ? docfile : doc)
+                  .replace(/%DOCFILE%/g, docfile)
+                  .replace(/%DIR%/g, docker ? './' : path.dirname(texPath).split(path.sep).join('/'))
+                  .replace(/%TMPDIR%/g, this.extension.builder.tmpDir)
     }
 
     get rootDir() {
         return path.dirname(this.rootFile)
     }
 
+    private rootFiles: { [key: string]: string } = {}
     get rootFile() {
```

Comment:
Please write a document on the difference between `rootFile` and `findRoot`. How can  we use them?

Reply:
Will add comments. In short, `rootFile` is an interface to `rootFiles`, which holds cache of roots w.r.t. workspaces. It is edited by the method `findRoot`.

Want to re-find the root file? -> `findRoot()`. It is costly so cannot invoke upon keystroke.
Want to just obtain the save root file location? -> `rootFile`